### ISSUE TITLE
Inaccurate complex tanh implementation

### ISSUE THREAD

**Issue by pv** [numba/numba#2922]
Numba complex tanh implementation returns nans for large real values, probably due to some intermediate under/overflow.

The following code fails:
```
import numpy as np
import numba

@numba.jit
def numba_tanh(x):
    return np.tanh(x)

points = [-700+1j, 700+1j]

for x in points:
    expected = np.tanh(x)
    got = numba_tanh(x)
    np.testing.assert_allclose(got, expected)
```
(tested on Numba 0.37.0, didn't test on 0.38.0 since llvmlite 0.23 fails to work on fedora 28)

**Comment by stuartarchibald** [comment]
Thanks for the report, I can reproduce.

**Comment by nleehone** [comment]
I just hit this problem in some code I was trying to speed up. The cmath.tanh trick works, but it's unclear why numpy would work differently under jit.

Here's a possibly simpler example of the problem.

```
import numpy as np
from numba import jit

def f(a):
    return np.tanh(a)

@jit
def g(a):
    return np.tanh(a)

print(f((1000-0j)))
# Gives (1+0j)

print(g((1000-0j)))
# Gives (nan+nanj)
```

Numpy version: '1.15.0'
Numba version: '0.39.0'

**Comment by seibert** [comment]
I did some digging, and it looks like due to the history of things being written at different times, we have two different implementations of `tanh` in Numba.

The implementation used when `np.tanh` is called on a complex dtype is here:

https://github.com/numba/numba/blob/c5461dd295663559b55f7e59280df9317062887c/numba/targets/npyfuncs.py#L1093

It uses raw llvmlite builder syntax, which we would generally try to avoid for numerical code in Numba today.

And the `cmath.tanh` implementation is here, adapted from what Python does:

https://github.com/numba/numba/blob/c04edc02deb40b84c3b21ed2a9835353181a5916/numba/targets/cmathimpl.py#L343-L367

Clearly the second implementation has better overflow behavior, but I haven't studied the first implementation enough yet to understand why it is different.

We should obviously unify these two implementations, presumably around the cmath version.

**Comment by nleehone** [comment]
Can we just re-use the code from cmathimpl.py in the npfuncs.py file? At the end of the day, we don't care how the underlying code works, as long as np.tanh gets translated into something that is correct and fast. 

It feels wrong to have two different versions of the tanh function. Maybe both should be abstracted out into another function that gets called by the cmath and np implementations?

I was looking at the rest of the code in npyfuncs.py and it looks like they are written in a similar way, and based off of the code in funcs.inc.src. I suspect that the other trig functions will also have problems with large complex values...

**Comment by seibert** [comment]
Yeah, there's no reason not to unify the two and use both.  In 2014-2016 Numba we sometimes tried too hard to mimic NumPy and Python separately, which I think leads to strange situations like this.
